src/schemas.jl

```diff
@@ -0,0 +1,662 @@
+
+where `RequiredFieldInfo` has the fields:
+
+- `name::Symbol`: the required field's name
+- `type::Union{Symbol,Expr}`: the required field's declared type constraint
+- `parameterize::Bool`: whether or not the required field is exposed as a parameter
+- `statement::Expr`: the required field's full assignment statement (as processed by `@version`, not necessarily as written)
+
+Note that `declaration` is primarily intended to be used for interactive discovery purposes, and
+does not include the contents of `declaration(parent(sv))`.
+"""
+declaration(sv::SchemaVersion) = throw(UnknownSchemaVersionError(sv))
+
+#####
+##### `SchemaVersion` printing
+#####
+
+Base.show(io::IO, sv::SchemaVersion) = print(io, "SchemaVersion(\"$(name(sv))\", $(version(sv)))")
+
+#####
+##### `SchemaVersion` Arrow (de)serialization
+#####
+
+const LEGOLAS_SCHEMA_VERSION_ARROW_NAME = Symbol("JuliaLang.Legolas.SchemaVersion")
+Arrow.ArrowTypes.arrowname(::Type{<:SchemaVersion}) = LEGOLAS_SCHEMA_VERSION_ARROW_NAME
+Arrow.ArrowTypes.ArrowType(::Type{<:SchemaVersion}) = String
+Arrow.ArrowTypes.toarrow(sv::SchemaVersion) = identifier(sv)
+Arrow.ArrowTypes.JuliaType(::Val{LEGOLAS_SCHEMA_VERSION_ARROW_NAME}, ::Any) = SchemaVersion
+Arrow.ArrowTypes.fromarrow(::Type{<:SchemaVersion}, id) = first(parse_identifier(id))
+
+#####
+##### `Tables.Schema` validation
+#####
+
+@inline accepted_field_type(::SchemaVersion, T) = T
+accepted_field_type(::SchemaVersion, ::Type{UUID}) = Union{UUID,UInt128}
+accepted_field_type(::SchemaVersion, ::Type{Symbol}) = Union{Symbol,String}
+
+"""
+    Legolas.find_violation(ts::Tables.Schema, sv::Legolas.SchemaVersion)
+
+For required field `f::F` of `sv`:
+
+- Define `A = Legolas.accepted_field_type(sv, F)`
+- If `f::T` is present in `ts`, ensure that `T <: A` or else immediately return `f::Symbol => T::DataType`.
+- If `f` isn't present in `ts`, ensure that `Missing <: A` or else immediately return `f::Symbol => missing::Missing`.
+
+Otherwise, return `nothing`.
+
+See also: [`Legolas.validate`](@ref), [`Legolas.complies_with`](@ref)
+"""
+find_violation(::Tables.Schema, sv::SchemaVersion) = throw(UnknownSchemaVersionError(sv))
+
+function _find_violation end
+
+"""
+    Legolas.validate(ts::Tables.Schema, sv::Legolas.SchemaVersion)
+
+Throws a descriptive `ArgumentError` if `!isnothing(find_violation(ts, sv))`,
+otherwise return `nothing`.
+
+See also: [`Legolas.find_violation`](@ref), [`Legolas.complies_with`](@ref)
+"""
+function validate(ts::Tables.Schema, sv::SchemaVersion)
+    result = find_violation(ts, sv)
+    isnothing(result) && return nothing
+    field, violation = result
+    ismissing(violation) && throw(ArgumentError("could not find expected field `$field` in $ts"))
+    expected = getfield(required_fields(sv), field)
+    throw(ArgumentError("field `$field` has unexpected type; expected <:$expected, found $violation"))
```

Comment:
would be nice if we could print all violations if there's more than 1

Reply:
Yeah. I'm fine leaving that to a subsequent release though. feel free to file an issue

Reply:
https://github.com/beacon-biosignals/Legolas.jl/issues/56